implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,344 @@
+        secure_channels: Arc<SecureChannels>,
+        addresses: Addresses,
+        identifier: IdentityIdentifier,
+        trust_policy: Arc<dyn TrustPolicy>,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+        credentials: Vec<Credential>,
+        trust_context: Option<TrustContext>,
+        remote_route: Option<Route>,
+        timeout: Option<Duration>,
+        role: Role,
+    ) -> Result<()> {
+        let vault = to_xx_vault(secure_channels.vault());
+        let identities = secure_channels.identities();
+        let state_machine: Box<dyn StateMachine> = if role.is_initiator() {
+            Box::new(
+                InitiatorStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        } else {
+            Box::new(
+                ResponderStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        };
+
+        let (mut callback_waiter, callback_sender) = ockam_node::callback::new_callback();
+
+        let worker = Self {
+            secure_channels,
+            callback_sender,
+            state_machine,
+            identifier: identifier.clone(),
+            role: role.clone(),
+            remote_route: remote_route.clone(),
+            addresses: addresses.clone(),
+            decryptor_handler: None,
+        };
+
+        WorkerBuilder::new(worker)
+            .with_mailboxes(Self::create_mailboxes(
+                &addresses,
+                decryptor_outgoing_access_control,
+            ))
+            .start(context)
+            .await?;
+
+        let decryptor_remote = addresses.decryptor_remote.clone();
+        debug!(
+            "Starting SecureChannel {} at remote: {}",
+            role.clone(),
+            &decryptor_remote
+        );
+        trace!(
+            "SecureChannel {} started for addresses: {:?}",
+            role.clone(),
+            addresses.clone(),
+        );
+
+        // before sending messages make sure that the handshake is finished and
+        // the encryptor worker is ready
+        if role.is_initiator() {
+            // wait until the handshake is finished
+            if let Some(timeout) = timeout {
+                callback_waiter.receive_timeout(timeout).await?;
+            } else {
+                callback_waiter.receive().await?;
+            }
+
+            // make sure that the encryptor has finished initializing before sending messages
+            context.wait_for(addresses.encryptor.clone()).await?;
```

Comment:
Sorry, but I still don't get why it's needed. What if we remove that line? What will happen?